src/util/logging/hooks.go

```diff
@@ -0,0 +1,86 @@
+package logging
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/sirupsen/logrus"
+)
+
+func NewReplayHook(logger *Logger) ReplayHook {
+	return ReplayHook{
+		Logger: logger,
+		excludeTypes: map[reflect.Type]int{
+			reflect.TypeOf(ReplayHook{}):    1,
+			reflect.TypeOf(ModuleLogHook{}): 1,
+		},
+	}
+}
+
+// Do not replay hooks of given exclude types
+func ExclusiveReplayHook(logger *Logger, exclude []reflect.Type) (h ReplayHook) {
```

Comment:
What does the replay hook do?

Reply:
It makes sure that hooks installed in the (default) top level logger are still executed for module-specific loggers . The later are just exact clones of the former also adding module-specific context. The separation is for implementing the same (`ExtendedFieldLogger`) interface for module loggers.

Hooks (if any) may be configured once for standard logger . By doing so all other module-specific loggers are immediate aware of the new configuration . This is expected behavior since the logging API does not provide access to the module-specific loggers , and it should not .

Type exclusion rules are needed so as to not to replay several module hooks (leading to log entries with wrong module tags) or another replay hook (e.g. leading to issues like invoking the hook chain twice, infinite loops , ...)

Reply:
`ExclusiveReplayHook` is not used anywhere?

Probably better to initialize with `[]interface{}` and do the reflection during construction, but if this isn't being used at all I'd remove it.